src/dhcp6relay/src/relay.cpp

```diff
@@ -0,0 +1,665 @@
+        current_position = tmp;
+        switch (ntohs(option->option_code)) {
+            case OPTION_RELAY_MSG:
+                memcpy(current_buffer_position, ((uint8_t *)option) + sizeof(struct dhcpv6_option), ntohs(option->option_length));
+                current_buffer_position += ntohs(option->option_length);
+                break;
+            default:
+                break;
+        }
+    }
+
+    strcpy(ifname, configs->interface.c_str());
+    memcpy(&target_addr.sin6_addr, &dhcp_relay_header->peer_address, sizeof(struct in6_addr));
+    target_addr.sin6_family = AF_INET6;
+    target_addr.sin6_flowinfo = 0;
+    target_addr.sin6_port = htons(546);
+    target_addr.sin6_scope_id = if_nametoindex(ifname);
+
+    send_udp(sock, buffer, target_addr, current_buffer_position - buffer);
+} 
+
+
+/**
+ * @code                callback(evutil_socket_t fd, short event, void *arg);
+ * 
+ * @brief               callback for libevent that is called everytime data is received at the filter socket
+ *
+ * @param fd            filter socket
+ * @param event         libevent triggered event  
+ * @param arg           callback argument provided by user
+ *
+ * @return              none
+ */
+void callback(evutil_socket_t fd, short event, void *arg) {
+    struct relay_config *config = (struct relay_config *)arg;
+    uint8_t message_buffer[4096];
+    uint32_t len = recv(config->filter, message_buffer, 4096, 0);
+    if (len <= 0) {
+        syslog(LOG_WARNING, "recv: Failed to receive data at filter socket\n");
+    }
+
+    char* ptr = (char *)message_buffer;
+    const uint8_t *current_position = (uint8_t *)ptr; 
+    const uint8_t *tmp = NULL;
+
+    auto ether_header = parse_ether_frame(current_position, &tmp);
+    current_position = tmp;
+
+    auto ip_header = parse_ip6_hdr(current_position, &tmp);
+    current_position = tmp;
+
+    if (ip_header->ip6_ctlun.ip6_un1.ip6_un1_nxt != IPPROTO_UDP) {
+        const struct ip6_ext *ext_header;
+        do {
+            ext_header = (const struct ip6_ext *)current_position;
+            current_position += ext_header->ip6e_len;
+        }
+        while (ext_header->ip6e_nxt != IPPROTO_UDP);
+    }  
+
+    auto udp_header = parse_udp(current_position, &tmp);
+    current_position = tmp;
+
+    auto dhcp_header = parse_dhcpv6_hdr(current_position);
+
+    if ((ntohs(udp_header->dest)) == 547 ) {
+        if(dhcp_header->msg_type == SOLICIT || REQUEST || CONFIRM || RENEW || REBIND || RELEASE || DECLINE) {
+            relay_client(config->local_sock, current_position, ntohs(udp_header->len) - sizeof(udphdr), ip_header, ether_header, config);
+        }    
+    }
+}
+
+/**
+ * @code                listen_server(void *arg);
+ * 
+ * @brief               listen to the DHCPv6 server
+ *
+ * @param arg           callback argument provided by user
+ *
+ * @return              none
+ */
+void listen_server(relay_config *config) {
+    while (serverListener) {
```

Comment:
why the server listener is not part of the even loop?

Reply:
Added server listener to libevent